src/expr.cpp

```diff
@@ -726,14 +728,66 @@ void ispc::InitSymbol(AddressInfo *ptrInfo, const Type *symType, Expr *initExpr,
                 return;
             }
 
+            // When we initialize uniform vector try to construct a vector from initializers and use a vector
+            // instruction for type conversion if needed. We can generate one instruction for type conversion when all
+            // initializers have the same atomic type. For example: const uniform float<4> f = {1.f, 2.f, 3.f, 4.f};
+            // const uniform double<4> d = {f[1], f[3], f.y, 0.0f};
+            // If it's not the case, or if initializer expression is not trivial (not of AtomicType), initialize vector
+            // per element.
+            if (symType->IsVectorType() && symType->IsUniformType() && nInits > 0) {
+                int nVectorElements = CastType<VectorType>(symType)->getVectorMemoryCount();
+                Expr *firstInit = exprList->exprs[0];
+                // Construct vector instruction only if initializer expression has AtomicType
+                if (CastType<AtomicType>(firstInit->GetType())) {
+                    // Get the type of first initializer expression
+                    const AtomicType *firstInitType = CastType<AtomicType>(firstInit->GetType());
+                    llvm::Type *llvmInitType = firstInitType->LLVMType(g->ctx);
+                    llvm::Value *initCast =
+                        llvm::UndefValue::get(llvm::FixedVectorType::get(llvmInitType, nVectorElements));
+                    bool isVectorIntializer = true;
+                    // Go through all initializer expressions and check if they are trivial and of the same type as
+                    // first expression
+                    for (int i = 0; i < nVectorElements; ++i) {
+                        if (i >= nInits) {
+                            // If we don't have enough initializer values, keep undef value
+                            continue;
+                        }
+                        // Check if initializer expression is trivial AtomicType
+                        if (CastType<AtomicType>(exprList->exprs[i]->GetType())) {
+                            llvm::Value *initializerValue = exprList->exprs[i]->GetValue(ctx);
+                            if (initializerValue == nullptr) {
+                                isVectorIntializer = false;
+                                break;
+                            }
+                            llvm::Type *initializerType = initializerValue->getType();
+                            // Check if initializer is of the same type as the first one
+                            if (initializerType != llvmInitType) {
+                                isVectorIntializer = false;
+                                break;
+                            }
+                            initCast = ctx->InsertInst(initCast, initializerValue, i);
```

Comment:
If this fails before completing the sequence, we would have a dead code, which is not good. I suggest collecting `Expr`s in the vector before starting the transformation and verifying that we will success. And as a second step doing the actual transformation.